proof/refine/ARM/Reply_R.thy

```diff
@@ -634,4 +618,486 @@ lemma replyRemoveTCB_invs':
+  apply (wpsimp wp: updateReply_list_refs_of_replies')
+  apply (subst arg_cong[where f=P])
+  prefer 2 apply assumption
+  apply (rule ext)
+  apply (frule ko_at'_replies_of')
+  by (auto split: if_split simp: map_set_def opt_map_def list_refs_of_reply'_def)
+
+lemma updateReply_obj_at'_inv:
+  "\<forall>x. P (f x) = P x \<Longrightarrow>
+   updateReply rPtr f \<lbrace>\<lambda>s. Q (obj_at' (P :: reply \<Rightarrow> bool) rp s)\<rbrace>"
+  apply (wpsimp wp: updateReply_wp_all)
+  apply (subgoal_tac "obj_at' P rp s = (obj_at' P rp (s\<lparr>ksPSpace := ksPSpace s(rPtr \<mapsto> KOReply (f ko))\<rparr>))")
+   apply simp
+  by (force simp: obj_at'_real_def ko_wp_at'_def objBitsKO_def ps_clear_def
+                  projectKO_reply)
+
+lemma updateReply_iflive'_weak:
+  "\<lbrace>if_live_then_nonz_cap' and ex_nonz_cap_to' replyPtr\<rbrace>
+   updateReply replyPtr f
+   \<lbrace>\<lambda>_. if_live_then_nonz_cap'\<rbrace>"
+  by (wpsimp wp: updateReply_iflive'_strong)
+
+lemma bindScReply_valid_objs'[wp]:
+  "\<lbrace>valid_objs' and reply_at' replyPtr and sc_at' scp\<rbrace>
+   bindScReply scp replyPtr
+   \<lbrace>\<lambda>_. valid_objs'\<rbrace>"
+  apply (solves wp | simp (no_asm_use) add: bindScReply_def split del: if_split cong: conj_cong |
+           wp hoare_when_wp haskell_assert_wp hoare_vcg_if_lift2 hoare_vcg_all_lift
+              hoare_vcg_disj_lift hoare_vcg_imp_lift' set_sc'.valid_objs'
+              set_reply'.valid_objs' setReply_obj_at' )+
+  apply (clarsimp simp: valid_obj'_def)
+  apply (drule obj_at_ko_at'[where p=scp], clarsimp)
+  apply (frule sc_ko_at_valid_objs_valid_sc', assumption)
+  apply (intro conjI impI allI;
+         clarsimp;
+         drule (1) ko_at'_inj;
+         clarsimp simp: valid_sched_context'_def valid_sched_context_size'_def objBits_def
+                        objBitsKO_def)
+  done
+
+crunches bindScReply
+  for valid_queues[wp]: valid_queues
+  and valid_queues'[wp]: valid_queues'
+  and valid_release_queue[wp]: valid_release_queue
+
+lemma bindScReply_sch_act_wf[wp]:
+  "bindScReply scPtr replyPtr \<lbrace>\<lambda>s. sch_act_wf (ksSchedulerAction s) s\<rbrace>"
+  unfolding bindScReply_def
+  by (wpsimp wp: sts_sch_act' hoare_vcg_all_lift hoare_vcg_if_lift hoare_drop_imps)
+
+lemma bindScReply_sym_refs_list_refs_of_replies':
+  "\<lbrace>\<lambda>s. sym_refs (list_refs_of_replies' s) \<and>
+        obj_at' (\<lambda>ko. replyPrev ko = None \<and> replyNext ko = None) replyPtr s
+       \<and> (\<forall>oldReplyPtr. obj_at' (\<lambda>ko. scReply ko = Some oldReplyPtr) scPtr s \<longrightarrow>
+           obj_at' (\<lambda>ko. replyNext ko = Some (Head scPtr)) oldReplyPtr s)\<rbrace>
+   bindScReply scPtr replyPtr
+   \<lbrace>\<lambda>_ s. sym_refs (list_refs_of_replies' s)\<rbrace>"
+  supply if_split [split del]
+  unfolding bindScReply_def
+  apply (wpsimp wp: hoare_vcg_if_lift2 updateReply_list_refs_of_replies'
```

Comment:
Indentation here, and in the following line. And the auto at the end doesn't need the `[1]`